Inset Mac OS traffic lights

haasal opened the pull request:
<!--
Update "[ ]" to "[x]" to check a box

Please make sure to read the Pull Request Guidelines: https://github.com/tauri-apps/tauri/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines
-->

### What kind of change does this PR introduce?
<!-- Check at least one. If you are introducing a new binding, you must reference an issue where this binding has been proposed, discussed and approved by the maintainers. -->

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [x] New Binding issue tauri-apps/tauri#4789
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?
<!-- If yes, please describe the impact and migration path for existing applications in an attached issue. -->

- [ ] Yes, and the changes were approved in issue #___
- [x] No

### Checklist
- [x] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [ ] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [x] I have added a convincing reason for adding this feature, if necessary

### Other information

**Important**: There still exist artifacts when resizing the window because the traffic lights are set just after the window has already been drawn.

haasal committed:
new example

haasal committed:
feat(macos): Fixed the traffic light example

Traffic lights get refreshed. Still artifacts visible

haasal committed:
feat(macos): Implemented and exposed a traffic light reposition method for Builder and Window

haasal committed:
fix(macos): fixed the issue with the artifact on resize

morajabi commented:
Thanks for this awesome PR! I had tried doing this only to fail at the "resize" artifacts: #494 Reading your solution made it clear. I should close that issue then.

PhotonQuantum commented:
This feature is fantastic!

However, I noticed a strange behavior: the hover effect still shows when I move the cursor to the "original" position of traffic lights. The electron implementation doesn't seem like have this behavior. I dug into its code but failed to find a solution.

Update: strangely, the problem goes away if I resize the window.

tr3ysmith commented:
Hey what's the status of this? I really would like to be able to apply inset on the traffic lights without work arounds

haasal commented:
Hi, me too. But I know really little about Tao. If this gets merged then we still have to implement it somehow into tauri.
Could someone review this please?

wusyong commented:
The implementation seems fine to me. But I'm not sure if it's good to merge now.
@amrbashir Do you think winit can accept this?
Btw you can get the raw window handle to change it yourself if you can't wait for it any longer.

amrbashir commented:
there is a good chance they may not accept but lets go ahead and merge this and if they didn't accept it upstream, we can implement it directly in tauri

amrbashir commented on examples/macos_traffic_lights.rs:
```diff
@@ -0,0 +1,54 @@
+// Copyright 2019-2021 Tauri Programme within The Commons Conservancy
```
example not needed

amrbashir commented on src/platform/macos.rs:
```diff
@@ -54,6 +54,8 @@ pub trait WindowExtMacOS {
   /// Sets whether or not the window has shadow.
   fn set_has_shadow(&self, has_shadow: bool);
+
+  fn set_traffic_light_inset(&self, inset: (f64, f64));
```
missing documentation and use `LogicalSize` instead of a tuple

amrbashir commented on src/platform/macos.rs:
```diff
@@ -338,6 +344,7 @@ pub trait WindowBuilderExtMacOS {
   fn with_resize_increments(self, increments: LogicalSize<f64>) -> WindowBuilder;
   fn with_disallow_hidpi(self, disallow_hidpi: bool) -> WindowBuilder;
   fn with_has_shadow(self, has_shadow: bool) -> WindowBuilder;
+  fn with_traffic_light_inset(self, inset: (f64, f64)) -> WindowBuilder;
```
missing documentation and use `LogicalSize` instead of a tuple

amrbashir commented on src/platform_impl/macos/window.rs:
```diff
@@ -134,6 +136,7 @@ unsafe fn create_view(
   })
 }
 
+// TODO: Modify this
```
why modify this?

haasal replied:
Honestly... I have no clue. I worked on this a long time ago

amrbashir commented on src/platform_impl/macos/window.rs:
```diff
@@ -262,6 +265,11 @@ fn create_window(
       }
 
+      if let Some((offset_x, offset_y)) = pl_attrs.traffic_light_inset {
+        println!("{} {}", offset_x, offset_y);
```
```suggestion
```

amrbashir commented on src/platform_impl/macos/window.rs:
```diff
@@ -1397,3 +1409,29 @@ unsafe fn set_max_inner_size<V: NSWindow + Copy>(window: V, mut max_size: Logica
   }
 }
+
+pub(crate) unsafe fn position_traffic_lights<V: NSWindow + Copy>(window: V, x: f64, y: f64) {
```
```suggestion
unsafe fn position_traffic_lights<W: NSWindow + Copy>(window: W, x: f64, y: f64) {
```

amrbashir commented on src/platform_impl/macos/window.rs:
```diff
@@ -1397,3 +1409,29 @@ unsafe fn set_max_inner_size<V: NSWindow + Copy>(window: V, mut max_size: Logica
+
+  let title_bar_container_view = close.superview().superview();
+
+  let close_rect: NSRect = msg_send![close, frame];
```
```suggestion
  let close_rect = close.frame();
```

amrbashir commented on src/platform_impl/macos/window.rs:
```diff
@@ -1397,3 +1409,29 @@ unsafe fn set_max_inner_size<V: NSWindow + Copy>(window: V, mut max_size: Logica
+  let button_height = close_rect.size.height;
+
+  let title_bar_frame_height = button_height + y;
+  let mut title_bar_rect = NSView::frame(title_bar_container_view);
```
```suggestion
  let mut title_bar_rect = title_bar_container_view.frame();
```

amrbashir commented on src/platform_impl/macos/window.rs:
```diff
@@ -1397,3 +1409,29 @@ unsafe fn set_max_inner_size<V: NSWindow + Copy>(window: V, mut max_size: Logica
+  let title_bar_frame_height = button_height + y;
+  let mut title_bar_rect = NSView::frame(title_bar_container_view);
+  title_bar_rect.size.height = title_bar_frame_height;
+  title_bar_rect.origin.y = NSWindow::frame(window).size.height - title_bar_frame_height;
```
```suggestion
  title_bar_rect.origin.y = window.frame().size.height - title_bar_frame_height;
```

amrbashir commented on src/platform_impl/macos/window.rs:
```diff
@@ -1397,3 +1409,29 @@ unsafe fn set_max_inner_size<V: NSWindow + Copy>(window: V, mut max_size: Logica
+  let _: () = msg_send![title_bar_container_view, setFrame: title_bar_rect];
+
+  let window_buttons = vec![close, miniaturize, zoom];
+  let space_between = NSView::frame(miniaturize).origin.x - NSView::frame(close).origin.x;
```
```suggestion
  let space_between = miniaturize.frame().origin.x - close_rect.origin.x;
```

amrbashir commented on src/platform_impl/macos/window.rs:
```diff
@@ -1397,3 +1409,29 @@ unsafe fn set_max_inner_size<V: NSWindow + Copy>(window: V, mut max_size: Logica
+  let title_bar_container_view = close.superview().superview();
+
+  let close_rect: NSRect = msg_send![close, frame];
+  let button_height = close_rect.size.height;
+
+  let title_bar_frame_height = button_height + y;
```
```suggestion
  let title_bar_frame_height = close_rect.size.height + y;
```

amrbashir commented on src/platform_impl/macos/window.rs:
```diff
@@ -1397,3 +1409,29 @@ unsafe fn set_max_inner_size<V: NSWindow + Copy>(window: V, mut max_size: Logica
+  let space_between = NSView::frame(miniaturize).origin.x - NSView::frame(close).origin.x;
+
+  for (i, button) in window_buttons.into_iter().enumerate() {
+    let mut rect: NSRect = NSView::frame(button);
```
```suggestion
    let mut rect = button.frame();
```

haasal committed:
Update src/platform_impl/macos/window.rs

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>
Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal committed:
Update src/platform_impl/macos/window.rs

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>
Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal committed:
Update src/platform_impl/macos/window.rs

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>
Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal committed:
Update src/platform_impl/macos/window.rs

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>
Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal committed:
Update src/platform_impl/macos/window.rs

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>
Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal committed:
Update src/platform_impl/macos/window.rs

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>
Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal committed:
Update src/platform_impl/macos/window.rs

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>
Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal committed:
Update src/platform_impl/macos/window.rs

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>
Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal commented:
I'm sorry this has taken so long. I didn't really have time to work on this. I will document the function and delete the example these days. I hope it's not to late to merge?

haasal committed:
Delete macos_traffic_lights.rs

Example not needed

Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal committed:
Update window.rs

Deleted outdated TODO

Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal committed:
refactor: Replaced type of inset with LogicalSize + docs

Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

amrbashir reviewed (commented):
Looks fine to me, however I'll hand this to @wusyong and @JonasKruckenberg (since you were interested in this)

amrbashir commented:
Please rebase the PR to fix the conflicts and add a change file

JonasKruckenberg commented:
I'll review this tomorrow, it's a super important imo and this or has been open wayy too long 😉 let's get this over the finish line

haasal commented:
Maybe it's important to add that I couldn't test the changes since my MacBook broke 🥺. If one of you could do that it would be amazing

JonasKruckenberg commented:
After testing this PR now, I feel `LogicalSize` makes no sense here, the way you currently call the method is as such

```rust
                                      // WTH means width here? and height?
.with_traffic_light_inset(LogicalSize { width: 10, height: 20 });
```

JonasKruckenberg commented:
Okay a few other observations:

1. While this does successfully inset the traffic lights, it does not increase the draggable area size
2. When resizing the window, it looses the inset

The second one really is a dealbreaker right now, I guess we need to implement a delegate and stuff just like electron does as well

JonasKruckenberg committed:
fixes

Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal commented:
> Okay a few other observations:
> 
>     1. While this does successfully inset the traffic lights, it does not increase the draggable area size
> 
>     2. When resizing the window, it looses the inset
> 
> 
> The second one really is a dealbreaker right now, I guess we need to implement a delegate and stuff just like electron does as well

I don't know if you fixed it already in that commit, but when I tested it with my supplied example the inset was persistent during resizing of the window.

JonasKruckenberg committed:
revert accidental commit

Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

JonasKruckenberg commented:
> I don't know if you fixed it already in that commit, but when I tested it with my supplied example the inset was persistent during resizing of the window.

Okay yes my bad, calling the `set_traffic_light_inset` method after the RedrawRequested event got fired fixes the artifacts.

My only question @haasal is why the `with_traffic_light_inset` method exists if it's not really doing what the user expects. This feels like misleading api design

JonasKruckenberg committed:
add back example

Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

JonasKruckenberg commented:
@amrbashir I added back the example, since it's indeed needed to use the method correctly (it's not enough to just call it once but it needs to be called after `RedrawRequested`)

haasal commented:
Yeah that was the reason I made the example. I already tried implementing this in Tauri itself using the raw window handle but I couldn't get rid of these artifacts. Do you think this method can now be used in tauri without causing artifacts?

haasal commented:
> My only question @haasal is why the `with_traffic_light_inset` method exists if it's not really doing what the user expects. This feels like misleading api design

I don't know how to implement it instead. This whole thing gave me big headaches because I couldn't find a way to set the inset without having to call the method in some weird way. If you have any idea how to do this without having to call the method after `RedrawRequested` it would be great!

amrbashir commented:
I still don't see why the example is needed since tao should already handle that internally and not require users to keep calling it in `RedrawRequested`

JonasKruckenberg commented:
> I still don't see why the example is needed since tao should already handle that internally and not require users to keep calling it in `RedrawRequested`

then we should change it to behave that way

haasal commented:
I really would like to get this over with but as said earlier I don't really know how. If you know how to directly inject something into the "window draw pipeline" (I.e. how to automatically call my inset-method/function without requiring the user to do some weird setup) I could take a look at it

amrbashir commented:
@haasal you need add an inset field on the `ViewState` struct here https://github.com/tauri-apps/tao/blob/c5d606dffeb1733ab06fd8c43eb3b9e7b2f553fe/src/platform_impl/macos/view.rs#L54 and your new function `set_traffic_light_inset` sets this field, then in `draw_rect` https://github.com/tauri-apps/tao/blob/c5d606dffeb1733ab06fd8c43eb3b9e7b2f553fe/src/platform_impl/macos/view.rs#L382  you have access to the state where you can read it and re-position the tarffic lights

haasal commented:
But the `with_traffic_light_inset()` method should remain as the public API method for the user to call, right?
The user just shouldn't ever have to call `set_traffic_light_inset()` manually.

amrbashir commented:
@haasal No, both methods should be public, both should modify the newly added state and triggers the traffic light positioning

haasal commented:
Sorry for not contributing to this in the last weeks. In spite of what I said I noticed that I don't currently have the time to finish this. I would really appreciate if one of you could push this over the finish line.

haasal committed:
fix(macos): traffic light offset now works as expected

The set_traffic_light_offset() method on window was removed and only the builder function now has to be called to set the offset. No more manual event handling is needed.

Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal committed:
refactor(macos): removed the traffic light example

the example is no longer needed as the api design is now clear and only one method call is needed to set the inset

Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal commented:
Got a new mac and fixed it. I removed the set_traffic_light_inset() method because it wasn't really necessary anymore and would add unnecessary complexity. The with_traffic_light_inset() method on WindowBuilder should now work as expected.

Edit: I added the `set_traffic_light_inset()` back in. I overestimated how complicated it would be to implement this

haasal committed:
feat(macos): added set_traffic_light_position back in

the set_traffic_light_position on window will now work as expected. I.e. only one call is necessary

Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal commented:
Imho apart from minor refactoring (if you see required) this should be ready to merge. Could you maybe take a look at this @JonasKruckenberg @amrbashir?

amrbashir commented on src/platform_impl/macos/view.rs:
```diff
@@ -414,13 +420,10 @@ extern "C" fn reset_cursor_rects(this: &Object, _sel: Sel) {
-          cursor:cursor
-      ];
-    }
+    let _: () = msg_send![this,
+        addCursorRect:bounds
+        cursor:cursor
+    ];
```
why was this check removed?

haasal replied:
Must have been on accident when deleting my own stuff

amrbashir commented on src/platform_impl/macos/window.rs:
```diff
@@ -1569,6 +1577,16 @@ impl WindowExtMacOS for UnownedWindow {
+
+  #[inline]
+  fn set_traffic_light_position(&self, position: (f64, f64)) {
+    // NOTE: Not sure if &self shouldn't be &mut self
```
```suggestion
```

amrbashir commented on src/platform_impl/macos/window.rs:
```diff
@@ -1569,6 +1577,16 @@ impl WindowExtMacOS for UnownedWindow {
   }
+
+  #[inline]
+  fn set_traffic_light_position(&self, position: (f64, f64)) {
```
```suggestion
  fn set_traffic_light_position<P: Into<Position>>(&self, position: P) {
```

haasal replied:
I don't really understand this. Do you want me to change only the public interface (i.e. set_traffic_..., with_...) or should I make the WindowBuilder field and the View attributes (or how they are called) also into a `position`? How do I extract x and y from this?

The x and y for the inset are LogicalPosition coordinates right? Can't we just use LogicalPosition as the inset Type instead of Position? This can be called passed using `LogicalPosition {x, y}` and so Jonas' argument against `LogicalSize` is satisfied

amrbashir replied:
This change should be added to all public APIs, and then you can convert them into logical position internally before applying the inset.

amrbashir replied:
Taking `Into<Position>` instead of `Into<LogicalPosition>` allows more flexibility and also to be consistent with other public APIs.

haasal replied:
The issue is that in the `set_traffic_light_position()` implementation I don't have access to the `self.window.scale_factor()` method for `WindowBuilder` which is needed to convert from `Position` to `LogicalPosition`

amrbashir replied:
you can get the scale factor using `NSWindow::backingScaleFactor(ns_window)` like this https://github.com/tauri-apps/tao/blob/bd8bafeca8a063351022cd5d5791f3111e8761d0/src/platform_impl/macos/window.rs#L820

haasal replied:
Resolved?

haasal committed:
delete comment

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>
Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal committed:
revert(macos): reverted accidental deletion of null check

Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal committed:
refactor(macos): changed traffic light inset to LogicalPosition<f64>

Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

haasal commented:
@amrbashir ok, I implemented your feedback. I hope the `LogicalPosition` is ok

haasal committed:
refactor(macos): changed pulic api back to Position

Also unified the api to always use inset instead of position

Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

amrbashir commented:
Thank you

amrbashir commented:
Just waiting for @wusyong to test and approve as well.

amrbashir commented:
@haasal looks like you didn't sign your commits and I can't merge the PR unless they are signed. You need to setup [commit signing](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits), then you can sign past commit like [this](https://superuser.com/questions/397149/can-you-gpg-sign-old-commits/1123928#1123928) for example.

haasal commented:
@amrbashir Do i also need to sign the partially verified commits? Also @JonasKruckenberg 's commits seem not to be signed either, right?

wusyong commented:
> Just waiting for @wusyong to test and approve as well.

I'm okay with the API but the title bar is expanded. Is this what you expect?

<img width="612" alt="Screenshot 2023-10-26 at 6 41 42 PM" src="https://github.com/tauri-apps/tao/assets/8409985/fcbc1be3-e828-4cbd-9550-80daa4c5f292">

haasal commented:
Oh god... I messed up didn't I?
@wusyong yes this is what one should expect. This will only work (look) properly with the overlay titlebar type

wusyong commented:
You can squash all your commits into one and then merge with dev. This should be the easiest way to sign commit.

haasal commented:
@wusyong I didn't do anything like this before and I don't want to mess up even worse. Do you mean pulling from upstream onto my dev branch and then merging with all of _my_ contributions squashed into one commit? Wouldn't this negate all my commit messages?

amrbashir committed:
cleanup after rebase

Signed-off-by: amrbashir <amr.bashir2015@gmail.com>

amrbashir committed:
more cleanup

amrbashir committed:
Merge branch 'dev' into inset

amrbashir committed:
fix macos

amrbashir reviewed (approved):
I guess it is ready for merge now, @wusyong would you mind testing and merge?
